tests/main/component-local-installs/task.yaml

```diff
@@ -0,0 +1,79 @@
+summary: Test installing components and snaps from local files.
+
+details: |
+  Test that we can install components and snaps from local files.
+  We test various combinations, like:
+    - installing a snap and some components for that snap
+    - installing a snap and some components for another snap (that is already installed)
+    - failing to install a snap and some components for another snap (that is not installed)
+    - installing multiple components for different snaps
+
+systems: [ubuntu-16.04-64, ubuntu-18.04-64, ubuntu-2*, ubuntu-core-*, fedora-*]
+
+execute: |
+  for container in test-snap-one/ test-snap-two/ comp-one/ comp-two/ comp-three/ comp-four/; do
+      snap pack $container
+  done
+
+  # install everything at once
+  snap install --dangerous \
+      ./test-snap-one_1.0_all.snap \
+      ./test-snap-two_1.0_all.snap \
+      ./test-snap-one+comp-one_1.0.comp \
+      ./test-snap-one+comp-two_1.0.comp \
+      ./test-snap-two+comp-three_1.0.comp \
+      ./test-snap-two+comp-four_1.0.comp
+
+  snap run test-snap-one comp-one
+  snap run test-snap-one comp-two
+  snap run test-snap-two comp-three
+  snap run test-snap-two comp-four
+
+  snap remove test-snap-one test-snap-two
+
+  # install only the snaps
+  snap install --dangerous \
+      ./test-snap-one_1.0_all.snap \
+      ./test-snap-two_1.0_all.snap
+  not snap run test-snap-one comp-one
+  not snap run test-snap-one comp-two
+  not snap run test-snap-two comp-three
+  not snap run test-snap-two comp-four
+
+  # install only the components
+  snap install --dangerous \
+      ./test-snap-one+comp-one_1.0.comp \
+      ./test-snap-one+comp-two_1.0.comp \
+      ./test-snap-two+comp-three_1.0.comp \
+      ./test-snap-two+comp-four_1.0.comp
+  snap run test-snap-one comp-one
+  snap run test-snap-one comp-two
+  snap run test-snap-two comp-three
+  snap run test-snap-two comp-four
+
+  snap remove test-snap-one test-snap-two
+
+  # install mixture of components for already installed snaps and a new snap
+  # with components
+  snap install --dangerous \
+    ./test-snap-one_1.0_all.snap \
+    ./test-snap-one+comp-one_1.0.comp
+  snap run test-snap-one comp-one
+  not snap run test-snap-one comp-two
+  snap install --dangerous \
+      ./test-snap-one+comp-two_1.0.comp \
+      ./test-snap-two_1.0_all.snap \
+      ./test-snap-two+comp-three_1.0.comp \
+      ./test-snap-two+comp-four_1.0.comp
+  snap run test-snap-one comp-one
+  snap run test-snap-one comp-two
+  snap run test-snap-two comp-three
+  snap run test-snap-two comp-four
+
+  snap remove test-snap-one test-snap-two
+
+  # fail to install a component for a snap that isn't installed
+  not snap install --dangerous \
+    ./test-snap-one_1.0_all.snap \
+    ./test-snap-one+comp-one_1.0.comp \
+    ./test-snap-two+comp-three_1.0.comp
```

Comment:
Should this check that one is installed but two is not after it finishes?

Also, it would be good to run the same command but with `--transaction=all-snaps`

Reply:
For this, the implementation checks for the missing "test-snap-two" prior to creating any tasks (or calling any of the snapstate functions), so the transaction has no impact on what is installed. In this case, none of the provided components/snaps will be installed.

Is that something we want to change?

Reply:
Does this mean that when we install more than one component simultaneously it is always a single transaction? Or is it just in this concrete case?

Thinking about this, it makes sense that it is a single transaction as the components are sort of an implementation detail of a single snap, contrary to snaps where each one is its own thing. cc @pedronis for his opinion on this.

Reply:
Yes, installing a snap and components will all happen in the same transaction..

Reply:
Getting back to this now, shouldn't this create two lanes, one for `snap-one` and component `snap-one+one`, and another for `snap-two+comp-three`? That is, one transaction per snap jointly with its components. And in this case only the for `snap-two` would fail, as we have not specified `--transaction-all`. The test should check that is the result.

And maybe worth adding another check (possibly before this one) using `--transaction-all` - in that case nothing should have been installed.

Reply:
I see what you mean now. However, this case is a little different, since failure here isn't caused by any of the tasks created to install the snap. This is more of a failed pre-condition check.

This is consistent with how we handle other similar failures. That is, we don't really consider lanes in situations like that.

Also, I was incorrect in my previous comment. This case we're looking at would create a new lane per-snap (and install the snap and its components in that lane). Again though, those lanes don't impact failures of this nature at the moment.

Reply:
Ok, it makes sense to fail on these pre-checks I guess. Maybe worth adding a comment expanding the explanation.